app/dashboard/utils.py

```diff
@@ -998,3 +1000,68 @@ def list_urls(lis, acc=None):
 
 def get_custom_avatars(profile):
     return CustomAvatar.objects.filter(profile=profile).order_by('-id')
+
+
+def _trim_null_address(address):
+    if address == '0x0000000000000000000000000000000000000000':
+        return '0x0'
+    else:
+        return address
+
+
+def _construct_transfer_filter_params(
+        recipient_address,
+        token_address,):
+
+    event_topic = keccak(text="Transfer(address,address,uint256)")
+
+    # [event, from, to]
+    topics = [
+        to_hex(event_topic),
+        None,
+        to_hex(encode_single('address', recipient_address))
+    ]
+
+    return {
+        "topics": topics,
+        "fromBlock": 0,
+        "toBlock": "latest",
+        "address": token_address,
+    }
+
+
+def _extract_sender_address_from_log(log):
+    return decode_single("address", log['topics'][1])
+
+
+def get_token_recipient_senders(network, recipient_address, token_address):
+    w3 = get_web3(network)
+
+    contract = w3.eth.contract(
+        address=token_address,
+        abi=erc20_abi,
+    )
+
+    balance = contract.functions.balanceOf(recipient_address).call()
+
+    if balance == 0:
+        return []
```

Comment:
I probably should take this out. If a user receives tokens, and then transfers them out; this function will return `[]` as if they never received them. I intended to create a fast exit path, but to determine if a address has ever received tokens, I dont think I can avoid retrieving the token transfer event logs.